### PR TITLE
Receive warning in Slack if fail queue is not empty (#1168)

### PR DESCRIPTION
**Pull request by GPelayo**
- Modified terraform files to include configurations for the new alarms.
- Modified the CWL_to_Slack secrets to make it possible for Cloudwatch to post on the azul-dev channel.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
@achave11, could you take a look? You are more familiar with CW alarms and metrics.

File: terraform/cloudwatch.tf.json.template.py [hunk 1]
```diff
@@ -134,6 +134,52 @@
...
+                },
+                "azul_fail_queue": {
+                    "alarm_name": f"{config.fail_queue_name}-message-count",
+                    "actions_enabled": True,
+                    "comparison_operator": "GreaterThanThreshold",
+                    "evaluation_periods": "1",
+                    "metric_name": "ApproximateNumberOfMessagesVisible",
+                    "namespace": "AWS/SQS",
+                    "period": "120",
+                    "statistic": "Average",
```
Comment (achave11) [inline terraform/cloudwatch.tf.json.template.py]: Although this works, we can be more strict towards the information we are collecting. 
```suggestion
                    "statistic": "SampleCount",
```

File: terraform/cloudwatch.tf.json.template.py [hunk 2]
```diff
@@ -134,6 +134,52 @@
...
+                },
+                "azul_fail_fifo_queue": {
+                    "alarm_name": f"{config.fail_fifo_queue_name}-message-count",
+                    "actions_enabled": True,
+                    "comparison_operator": "GreaterThanThreshold",
+                    "evaluation_periods": "1",
+                    "metric_name": "ApproximateNumberOfMessagesVisible",
+                    "namespace": "AWS/SQS",
+                    "period": "120",
+                    "statistic": "Average",
```
Comment (achave11) [inline terraform/cloudwatch.tf.json.template.py]: Same as above.

File: terraform/cloudwatch.tf.json.template.py (hunk 1 above)
Comment (GPelayo) [reply terraform/cloudwatch.tf.json.template.py]: According to the [Amazon docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html), `SampleCount` is just the number times a metric has been measured within a certain period. If you wanted a stricter statistic, wouldn't `Sum` be better?

File: terraform/cloudwatch.tf.json.template.py [hunk 3]
```diff
@@ -134,9 +134,33 @@
...
+                },
+                **{
+                    f"{queue}-queue": {
+                        "alarm_name": f"{queue}-message-count",
+                        "actions_enabled": True,
+                        "comparison_operator": "GreaterThanThreshold",
+                        "evaluation_periods": "1",
+                        "metric_name": "ApproximateNumberOfMessagesVisible",
+                        "namespace": "AWS/SQS",
+                        "period": "300",
```
Comment (hannes-ucsc) [inline terraform/cloudwatch.tf.json.template.py]:
```suggestion
                        "period": "300",  # SQS pushes metrics at most every 5 min, lower periods wouldn't make sense
```

**Comment by hannes-ucsc** [comment]
Before requestiing the next review cycle, please provide proof of how this was tested, @GPelayo.

**Comment by GPelayo** [comment]
I test these alarms by first commenting out all the shared resources to prevent those resources being overwritten when I deploy the alarms in `cloudwatch.tf.json.template.py`.
Then, I inverse the condition in the expression, `if config.enable_monitoring`, to enable the alarms without modifying the environment settings. Below is `cloudwatch.tf.json.template.py` with the above changes.

```
emit({
    "resource": [
        # *([] if config.share_es_domain else [{
        #     "aws_cloudwatch_metric_alarm": {
        #         "CPUUtilization": {
        #             "alarm_name": config.es_domain+"-CPUUtilization",
        #             "actions_enabled": True,
        #             "comparison_operator": "GreaterThanOrEqualToThreshold",
        #             "evaluation_periods": "2",
        #             "metric_name": "CPUUtilization",
        #             "namespace": "AWS/ES",
        #             "period": "3600",
        #             "statistic": "Average",
        #             "threshold": "85",
        #             "alarm_description": json.dumps({
        #                 "slack_channel": "dcp-ops-alerts",
        #                 "description": config.es_domain + " CPUUtilization alarm"
        #             }),
        #             "dimensions": {
        #                 "ClientId": aws.account,
        #                 "DomainName": config.es_domain
        #             },
        #             "alarm_actions": [
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
        #             ]
        #         }
        #     }
        # },
        # {
        #     "aws_cloudwatch_metric_alarm": {
        #         "FreeStorageSpace": {
        #             "alarm_name": config.es_domain + "-FreeStorageSpace",
        #             "actions_enabled": True,
        #             "comparison_operator": "LessThanOrEqualToThreshold",
        #             "evaluation_periods": "1",
        #             "metric_name": "FreeStorageSpace",
        #             "namespace": "AWS/ES",
        #             "period": "300",
        #             "statistic": "Average",
        #             "threshold": "14000",
        #             "alarm_description": json.dumps({
        #                 "slack_channel": "dcp-ops-alerts",
        #                 "description": config.es_domain + " FreeStorageSpace alarm"
        #             }),
        #             "dimensions": {
        #                 "ClientId": aws.account,
        #                 "DomainName": config.es_domain
        #             },
        #             "alarm_actions": [
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
        #             ]
        #         }
        #     }
        # },
        # {
        #     "aws_cloudwatch_metric_alarm": {
        #         "JVMMemoryPressure": {
        #             "alarm_name": config.es_domain + "-JVMMemoryPressure",
        #             "actions_enabled": True,
        #             "comparison_operator": "GreaterThanOrEqualToThreshold",
        #             "evaluation_periods": "1",
        #             "metric_name": "JVMMemoryPressure",
        #             "namespace": "AWS/ES",
        #             "period": "300",
        #             "statistic": "Minimum",
        #             "threshold": "65",
        #             "alarm_description": json.dumps({
        #                 "slack_channel": "dcp-ops-alerts",
        #                 "description": config.es_domain + " JVMMemoryPressure alarm"
        #             }),
        #             "dimensions": {
        #                 "ClientId": aws.account,
        #                 "DomainName": config.es_domain
        #             },
        #             "alarm_actions": [
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
        #                 f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
        #             ]
        #         }
        #     }
        # }]),
        {
            "aws_cloudwatch_metric_alarm": {
                "azul_health": {
                    "alarm_name": f"azul-{config.deployment_stage}",
                    "actions_enabled": True,
                    "comparison_operator": "LessThanThreshold",
                    "evaluation_periods": "1",
                    "metric_name": "HealthCheckStatus",
                    "namespace": "AWS/Route53",
                    "period": "120",
                    "statistic": "Minimum",
                    "threshold": "1.0",
                    "alarm_description": json.dumps({
                        "slack_channel": "azul-dev",
                        "environment": config.deployment_stage,
                        "description": f"azul-{config.deployment_stage} HealthCheckStatus alarm"
                    }),
                    "alarm_actions": [
                        f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
                        f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
                    ],
                    "dimensions": {
                        "HealthCheckId": "${aws_route53_health_check.composite-azul.id}",
                    }
                },
                "data_portal_health": {
                    "alarm_name": f"data-browser-{config.deployment_stage}",
                    "actions_enabled": True,
                    "comparison_operator": "LessThanThreshold",
                    "evaluation_periods": "1",
                    "metric_name": "HealthCheckStatus",
                    "namespace": "AWS/Route53",
                    "period": "120",
                    "statistic": "Minimum",
                    "threshold": "1.0",
                    "alarm_description": json.dumps({
                        "slack_channel": "data-browser",
                        "environment": config.deployment_stage,
                        "description": f"data-browser-{config.deployment_stage} HealthCheckStatus alarm"
                    }),
                    "alarm_actions": [
                        f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
                        f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
                    ],
                    "dimensions": {
                        "HealthCheckId": "${aws_route53_health_check.composite-portal.id}",
                    }
                },
                **{
                    f"{queue.replace('.', '-')}-queue": {
                        "alarm_name": f"{queue}-message-count",
                        "actions_enabled": True,
                        "comparison_operator": "GreaterThanThreshold",
                        "evaluation_periods": "1",
                        "metric_name": "ApproximateNumberOfMessagesVisible",
                        "namespace": "AWS/SQS",
                        "period": "300",  # SQS pushes metrics at most every 5 min, lower periods wouldn't make sense
                        "statistic": "Maximum",
                        "threshold": "0.0",
                        "alarm_description": json.dumps({
                            "slack_channel": "azul-dev",
                            "environment": config.deployment_stage,
                            "description": f"{queue} ApproximateNumberOfMessagesVisible alarm"
                        }),
                        "dimensions": {
                            "QueueName": queue
                        },
                        "alarm_actions": [
                            f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
                            f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
                        ],
                    } for queue in (config.fail_queue_name, config.fail_fifo_queue_name)
                }
            }
        }
    ]
} if not config.enable_monitoring else None)
```

The if statements in `route53_health_check.tf.json.template.py` that include the `config.enable_monitoring` are also inverted, since the `azul_health` and `data_portal_health` alarms are dependant on that template.

Finally, I run `make terraform` to deploy the alarms.
Once alarms are deployed, I check in the AWS Cloudwatch console if the alarms appear in CloudWatch without an `INSUFFICIENT` state.

File: terraform/cloudwatch.tf.json.template.py [hunk 4]
```diff
@@ -134,6 +134,31 @@
                     "dimensions": {
                         "HealthCheckId": "${aws_route53_health_check.composite-portal.id}",
                     }
+                },
+                **{
+                    f"{queue.replace('.', '-')}-queue": {
```
Comment (hannes-ucsc) [inline terraform/cloudwatch.tf.json.template.py]:
```suggestion
                    f"{queue.replace('.', '_')}_queue": {
```

**Pull request merged**